# Viikko 6 - Django

## a) CRM-ohjelma

Ensimmäisessä kohdassa piti luoda yksinkertainen ohjelma Djangolla. Lähdin tekemään samanlaista CRM-ohjelmaa kuin Terolla. [Lähde](https://terokarvinen.com/2022/django-instant-crm-tutorial/)

Käynnistin virtuaalikoneeni ja aloitin komennolla sudo apt-get update. (Kello 18.12)

Seuraavana piti asentaa kehitysympäristö sudo apt-get -y install virtualenv

Ja tämän jälkeen virtualenv --system-site-packages -p python3 env/

![image](https://github.com/user-attachments/assets/598b7754-70b8-4184-91c5-986081ee265c)

Seuraavaksi otettiin virtuaaliympäristö käyttöön source env/bin/activate. Komentokehote näyttikin alkuun (env)

![image](https://github.com/user-attachments/assets/9fa13e8b-65aa-477d-a6d0-ae3b5126b9bb)

Seuraavaksi asensin micron (sudo apt-get install micro) ja loin tekstitiedoston requirements.txt, jonne kirjoitin django.

![image](https://github.com/user-attachments/assets/bf829d38-49dc-4e7a-96a6-0fb263a22792)

![image](https://github.com/user-attachments/assets/24e6c277-feb0-4a45-9e6c-65060a0337e9)

Seuraavaksi asennettiin django käyttämällä pip-komentoa. 

![image](https://github.com/user-attachments/assets/7797a3d7-5646-4e6b-a55f-4e45064fd6be)

Tarkistin, että sain uusimman version djangosta.

![image](https://github.com/user-attachments/assets/b35060bb-0c41-4c6f-871c-00098bc7eadb)

Asennuksen jälkeen olikin aika aloittaa projekti. Aloitin projektin ajamalla django-admin startproject projektinnimi. Valitsin Teron teroco-tyyliin eerogrp. Sitten cd eerogrp ja ./manage.py runserver.

![image](https://github.com/user-attachments/assets/5498ac6b-2039-4369-808f-4aff24e954ea)

Asennus onnistui.

![image](https://github.com/user-attachments/assets/d87b4af7-f044-42d4-b1ff-3b1db6d6ca9d)

Seuraavaksi lisäsin käyttäjän eero, jolle loin vahvan salasana pwgen avulla. 

![image](https://github.com/user-attachments/assets/eeb6a453-419a-4403-9836-1b1940a407b7)

Kokeilin kirjautua sisään, mutta näkymä oli tämä.

![image](https://github.com/user-attachments/assets/02ad20f6-9296-4174-bca6-d7dc918c336f)

Tajusin, että olin sulkenut serverin. Joten käynnistin sen uudestaan ./manage.py runserver. Nyt pääsin kirjautumaan admin-sivulta.

![image](https://github.com/user-attachments/assets/0cde2fc8-5973-4a7c-a717-64cc8bbb01a8)

Tässä vaiheessa siirryin tauolle. (Kello 18:43) Hommat jatkui kello 18:55.

Seuraavaksi lähdin luomaan tietokantaa asiakkaille ./manage.py startapp crm

Ja micro eerogrp/settings.py

![image](https://github.com/user-attachments/assets/0c995e58-4d8a-4be8-b270-1ee49000ee80)

![image](https://github.com/user-attachments/assets/477f83f5-d6e7-44d4-a3f8-eabbe77ff0b8)

Lisäsin Installed Apps listaan crm. Ctrl+S ja Ctrl+Q tallensi ja poistui microsta.

![image](https://github.com/user-attachments/assets/ef4a2b99-0983-422d-8d1d-22830fd2983d)

Seuraavaksi lisäsin malleja crm:ään komennolla micro crm/models.py. En saanut kopioitua tekstiä suoraan Terolta, joten kirjoitin käsin vastaavan tekstin. 

![image](https://github.com/user-attachments/assets/8ff41f33-1474-44b9-9bf3-9fb8f8198e67)

Seuraavaksi ajoin esimerkin mukaisesti ./manage.py makemigrations ja ./manage.py migrate

![image](https://github.com/user-attachments/assets/6cf12323-2ec1-4237-b1d6-c467ac59d269)

Sitten ajoin micro crm/admin.py ja lisäsin seuraavan tekstin. Tässä kohdassa rupesin miettimään, onko tässä kyseessä käyttäjänimi "admin" vai admin. Lähdin kuitenkin kokeilemaan ensin tällä tapaa, ja jos tämä ei toimi niin tiedän tehdä saman omalla käyttäjänimelläni.

![image](https://github.com/user-attachments/assets/58bc9a58-0f64-44c4-a1cf-9a19a0dde454)

Käynnistin serverin uudestaan ja kirjauduin sisään. Customers-näkymä löytyi ja lisäsin kolme asiakasta.

![image](https://github.com/user-attachments/assets/9dfeaed3-cd01-4b35-b199-456ef15e8838)

Poistin lisäämäni asiakkaat samalla.

![image](https://github.com/user-attachments/assets/754ed725-9d0f-4446-9f09-5b102900488c)

Muokkasin vielä microssa crm/models.py, jotta asiakkaiden nimet näkyisivät oikein.

![image](https://github.com/user-attachments/assets/1d7864cf-2423-4d2b-a51c-3f99875cf5b6)

Lisäsin asiakkaat ja nimet näkyivät oikein.

![image](https://github.com/user-attachments/assets/4c6df9e6-9247-46a3-b337-e0aadc38c142)

Tässä kohtaa olikin taas aika tauolle. (Kello 19:16)

## b) Tuotantotyyppinen asennus

Aluksi muokkasin verkkosivua hieman. (aloitus kello 19:30)

![image](https://github.com/user-attachments/assets/0128c400-fd1a-4f8f-9c14-5f6089e69e50)

Lisäsin virtual hostin sudoedit /etc/apache2/sites-available/eerogrp.conf ja lisäsin seuraavan tekstin.

![image](https://github.com/user-attachments/assets/b659bdbc-bc30-45b5-b7d4-b14705878ffb)

Sitten täytyi laittaa äsken tehty sivu oletussivuksi. 

sudo a2ensite eerogrp.conf ja sudo a2dissite 000-default.conf.

sudo systemctl restart apache2

Sitten tarkistin, että äskeinen konfiguraatio toimii /sbin/apache2ctl configtest. Kokeilin myös curl http://localhost/static. Välissä oli myös sudo systemctl restart apache2

![image](https://github.com/user-attachments/assets/597e39a6-0c25-4ec1-847a-9b2d35fcaf20)

(ruokatauko @ 19:40) Illallisen ja koiran ulkoilutuksen jälkeen hommat jatkui 20:40.

Virtualenv oli jo asennettuna, joten laitoin sen kansioon /publicwsgi.

![image](https://github.com/user-attachments/assets/0631a1d8-99d3-4c97-8c72-92526e343771)

Toistin samat vaiheet kuin kohdassa a):

![image](https://github.com/user-attachments/assets/f5cee0f9-ef7f-4b2d-bec3-c7bd00068b5a)

Django asennettuna aloitin uuden projektin django-admin startproject eeroco. (tässä kohtaa myös huomasin, että olin tehnyt toisella nimellä alkuun ja kävin muuttamassa)

Muutin Virtualhost config filen halutun näköiseksi:

![image](https://github.com/user-attachments/assets/d49771e0-0ba9-44f0-ba4f-dec172f68d21)

Tässä kohtaa alkoivat ongelmat. Localhost näytti vain Apachen oletussivua, eikä config filen muuttaminen auttanut. Pyörittelin eri vaihtoehtoja ja tarkistin kirjoitusvirheitä.

Lopulta keksin tarkistaa polut. Ensinnäkin löysin että piti käyttää python3.11 eikä python3.9. Viimeinen virhe taisi olla puuttuva / TDIR kohdassa. Lopulta sain toivotun vastauksen curl -s localhost|grep title.

![image](https://github.com/user-attachments/assets/2d0782da-cdea-4322-89e7-91375d8dea92)

Jatkoin tästä eteenpäin muokkaamaan asetuksia.

![image](https://github.com/user-attachments/assets/c28a514b-559e-4486-9e0e-410bb3c0f0a5)

Muokkasin asetuksia:

![image](https://github.com/user-attachments/assets/763e0eeb-1878-4dc5-b6a7-a58348c15272)

Pääsin kirjautumaan http://localhost/admin, ja sivu näytti tältä:

![image](https://github.com/user-attachments/assets/30f45dda-152a-492c-99e4-17552d867195)

Ohjeiden perusteella luin, ettei siellä pitäisikään vielä näkyä mitään. Lähdin siis muokkaamaan /settings.py tiedostoa uudestaan.

Lisäsin ohjeen mukaisesti import os
STATIC_ROOT = os.path.join(BASE_DIR, 'static/') ja sitten ajoin ./manage.py collectstatic

![image](https://github.com/user-attachments/assets/735d82da-451d-4755-a745-05dcd7ad566b)

Kokeilin miltä localhost/admin näyttää:

![image](https://github.com/user-attachments/assets/8dea5076-0759-4ef3-98bc-7f9085263d67)

Kirjautumisen jälkeen sain kuitenkin ainoastaan saman Server Error (500)

Tämän jälkeen yritin etsiä [Django Admin Server Error 500](https://stackoverflow.com/questions/41667222/server-error-500-when-trying-to-log-in-in-django-admin-page)

Yritin myös etsiä lokeista virheitä, löysin vain tämän AH00492: caught SIGWINCH, shutting down gracefully. Google-haulla en löytänyt apua.

Kokeilin vaihdella DEBUG = True, mutta tulosta ei tullut. ./manage.py runserver antoi tämän tuloksen:

![image](https://github.com/user-attachments/assets/d36ab035-058c-44b1-80fc-34908f7604d1)

Kokeilin lisätä "127.0.0.1" Allowed hosts kohtaan, mutta muutosta ei tapahtunut.

Lisää kikkailua DEBUG kanssa, mutta konsoli ei näytä ongelmaa missään lokissa. Tässä vaiheessa täytyy luovuttaa ja kysyä tunnilla ratkaisua.































